File: server/kafka-backend/src/infrastructure/services/banking.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectDataSource } from '@nestjs/typeorm';
import { DataSource } from 'typeorm';
import { AccountOrmEntity } from '../adapters/out/persistence/typeorm/entities/account.orm-entity';
import { CardOrmEntity } from '../adapters/out/persistence/typeorm/entities/card.orm-entity';
import { CryptoService } from './crypto.service';

const BANK_CODE = 'INTB';
const IBAN_ACCOUNT_LENGTH = 16;
const CARD_BIN = '499999';
const CARD_LENGTH = 16;
const DEFAULT_CARD_LIFETIME_YEARS = 3;
const MAX_RETRIES = 7;

function generateRandomDigits(length: number): string
{
    let result = '';
    for(let i = 0; i < length; i++)
    {
        result += Math.floor(Math.random() * 10).toString();
    }
    return result;
}

function computeIBANChecksum(bban: string): string
{
    const numeric = [...(bban + 'RO00')].map((c) =>
        /[A-Z]/.test(c) ? (c.charCodeAt(0) - 55).toString() : c,
    ).join('');
    const mod = BigInt(numeric) % 97n;
    const checksum = 98n - mod;
    return checksum.toString().padStart(2, '0');
}

function generateIBAN(currency: string, countryCode: string): string
{
    const country = countryCode.toUpperCase().slice(0, 2);
    const bban =
        BANK_CODE +
        generateRandomDigits(4) +
        (currency === 'RON' ? 'RON' : generateRandomDigits(3)) +
        generateRandomDigits(IBAN_ACCOUNT_LENGTH - BANK_CODE.length - 4 - (currency === 'RON' ? 3 : 3));
    const checksum = computeIBANChecksum(bban);
    return `${country}${checksum}${bban}`;
}

@Injectable()
export class BankingService
{
    private readonly logger = new Logger(BankingService.name);

    constructor(
        @InjectDataSource()
        private readonly dataSource: DataSource,
        private readonly cryptoService: CryptoService,
    ) {}

    async createAccountAndCard(
        userId: number,
        currency: string,
        countryCode: string,
    ): Promise<{ account: any; card: any }>
    {
        const IBAN = generateIBAN(currency, countryCode);

        return await this.dataSource.transaction(async (entityManager) =>
        {
            const accountRepo = entityManager.getRepository(AccountOrmEntity);
            const cardRepo = entityManager.getRepository(CardOrmEntity);

            const account = accountRepo.create(
            {
                userId,
                IBAN,
                moneda: currency,
                sold: 0,
            });
            const savedAccount = await accountRepo.save(account);
            const accountId = savedAccount.id;

            const cardNumber = CARD_BIN + generateRandomDigits(CARD_LENGTH - CARD_BIN.length);
            const cvv = generateRandomDigits(3);
            const now = new Date();
            const expiry = new Date(now.getFullYear() + DEFAULT_CARD_LIFETIME_YEARS, now.getMonth(), 1);
            const expiryMMYY = `${(expiry.getMonth() + 1).toString().padStart(2, '0')}/${expiry.getFullYear().toString().slice(-2)}`;

            const encryptedCard = this.cryptoService.encryptAESGCM(cardNumber);
            const encryptedCVV = this.cryptoService.encryptAESGCM(cvv);
            const encryptedExpiry = this.cryptoService.encryptAESGCM(expiryMMYY);
            const token = `tok_${generateRandomDigits(16)}`;

            const card = cardRepo.create(
            {
                userId,
                numarCard: encryptedCard,
                cvv: encryptedCVV,
                dataExpirare: encryptedExpiry,
                detinator: '',
                token,
                accountId,
            });
            const savedCard = await cardRepo.save(card);

            this.logger.log(
                { userId, accountId, IBAN, currency },
                'Account and card created',
            );

            return {
                account:
                {
                    id: accountId,
                    IBAN: savedAccount.IBAN,
                    moneda: savedAccount.moneda,
                    sold: savedAccount.sold,
                },
                card:
                {
                    id: savedCard.id,
                    token,
                    last4: cardNumber.slice(-4),
                    expiryMMYY,
                    accountId,
                },
            };
        });
    }

    async createAccountAndCardWithRetry(
        userId: number,
        currency: string,
        countryCode: string,
        retries: number = MAX_RETRIES,
    ): Promise<{ account: any; card: any }>
    {
        for(let attempt = 1; attempt <= retries; attempt++)
        {
            try
            {
                return await this.createAccountAndCard(userId, currency, countryCode);
            }
            catch (err: any)
            {
                const isUniqueViolation = err.code === '23505';
                if(isUniqueViolation && attempt < retries)
                {
                    this.logger.warn({ attempt }, 'IBAN collision, retrying...');
                    continue;
                }
                this.logger.error(err, 'createAccountAndCard failed');
                throw err;
            }
        }
        throw new Error('Failed to generate unique IBAN after multiple attempts');
    }
}

File: server/backend/src/main/java/com/intbank/service/BankingService.java
package com.intbank.service;

import com.intbank.infrastructure.persistence.entity.AccountJpaEntity;
import com.intbank.infrastructure.persistence.entity.CardJpaEntity;
import com.intbank.infrastructure.persistence.entity.UserJpaEntity;
import com.intbank.infrastructure.persistence.repository.AccountJpaRepository;
import com.intbank.infrastructure.persistence.repository.CardJpaRepository;
import com.intbank.infrastructure.persistence.repository.UserJpaRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.Map;
import java.util.Random;

@Service
public class BankingService
{

    private static final Logger log = LoggerFactory.getLogger(BankingService.class);
    private static final String BANK_CODE = "INTB";
    private static final int IBAN_ACCOUNT_LENGTH = 16;
    private static final String CARD_BIN = "499999";
    private static final int CARD_LENGTH = 16;
    private static final int DEFAULT_CARD_LIFETIME_YEARS = 3;
    private static final int MAX_RETRIES = 7;
    private static final Random RANDOM = new Random();

    private final AccountJpaRepository accountRepo;
    private final CardJpaRepository cardRepo;
    private final UserJpaRepository userRepo;
    private final CryptoService cryptoService;

    public BankingService(
        AccountJpaRepository accountRepo,
        CardJpaRepository cardRepo,
        UserJpaRepository userRepo,
        CryptoService cryptoService
    )
    {
        this.accountRepo = accountRepo;
        this.cardRepo = cardRepo;
        this.userRepo = userRepo;
        this.cryptoService = cryptoService;
    }

    @Transactional
    public Map<String, Object> createAccountAndCard(long userId, String currency, String countryCode)
    {
        String iban = generateIBAN(currency, countryCode);

        UserJpaEntity user = userRepo.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));

        AccountJpaEntity account = new AccountJpaEntity();
        account.setUser(user);
        account.setIBAN(iban);
        account.setMoneda(currency);
        account.setSold(BigDecimal.ZERO);
        account = accountRepo.save(account);

        Long accountId = account.getId();
        String cardNumber = CARD_BIN + generateRandomDigits(CARD_LENGTH - CARD_BIN.length());
        String cvv = generateRandomDigits(3);

        YearMonth expiry = YearMonth.now().plusYears(DEFAULT_CARD_LIFETIME_YEARS);
        String expiryMMYY = String.format("%02d/%02d", expiry.getMonthValue(), expiry.getYear() % 100);

        String encryptedCard = cryptoService.encryptAESGCM(cardNumber);
        String encryptedCVV = cryptoService.encryptAESGCM(cvv);
        String encryptedExpiry = cryptoService.encryptAESGCM(expiryMMYY);
        String token = "tok_" + generateRandomDigits(16);

        CardJpaEntity card = new CardJpaEntity();
        card.setUser(user);
        card.setAccount(account);
        card.setNumarCard(encryptedCard);
        card.setCvv(encryptedCVV);
        card.setDataExpirare(encryptedExpiry);
        card.setDetinator(user.getNume() + " " + user.getPrenume());
        card.setToken(token);
        card = cardRepo.save(card);

        log.info("Account and card created: userId={}, accountId={}, IBAN={}, currency={}", userId, accountId, iban, currency);

        return Map.of(
                "account", Map.of("id", accountId, "IBAN", iban, "moneda", currency, "sold", 0),
                "card", Map.of("id", card.getId(), "token", token, "last4", cardNumber.substring(cardNumber.length() - 4),
                        "expiryMMYY", expiryMMYY, "accountId", accountId)
        );
    }

    public Map<String, Object> createAccountAndCardWithRetry(long userId, String currency, String countryCode)
    {
        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++)
        {
            try
            {
                return createAccountAndCard(userId, currency, countryCode);
            }
            catch (Exception err)
            {
                if (err.getMessage() != null && err.getMessage().contains("unique") && attempt < MAX_RETRIES)
                {
                    log.warn("IBAN collision on attempt {}, retrying...", attempt);
                    continue;
                }
                log.error("createAccountAndCard failed", err);
                throw err;
            }
        }
        throw new RuntimeException("Failed to generate unique IBAN after " + MAX_RETRIES + " attempts");
    }

    private String generateIBAN(String currency, String countryCode)
    {
        String country = countryCode.toUpperCase().substring(0, 2);
        String bban = BANK_CODE + generateRandomDigits(4) + (currency.equals("RON") ? "RON" : generateRandomDigits(3))
                + generateRandomDigits(IBAN_ACCOUNT_LENGTH - BANK_CODE.length() - 4 - 3);
        String checksum = computeIBANChecksum(bban);
        return country + checksum + bban;
    }

    private String computeIBANChecksum(String bban)
    {
        StringBuilder numeric = new StringBuilder();
        for (char c : (bban + "RO00").toCharArray())
        {
            if (c >= 'A' && c <= 'Z')
            {
                numeric.append(c - 'A' + 10);
            }
            else
            {
                numeric.append(c);
            }
        }
        java.math.BigInteger mod = new java.math.BigInteger(numeric.toString()).mod(java.math.BigInteger.valueOf(97));
        long checksum = 98 - mod.longValue();
        return String.format("%02d", checksum);
    }

    private String generateRandomDigits(int length)
    {
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++)
        {
            sb.append(RANDOM.nextInt(10));
        }
        return sb.toString();
    }
}